### PR TITLE
trivial-builders: rework runCommand' into runCommandWith && expose 

### PR DESCRIPTION
**Pull request by sternenseemann**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sternenseemann** [comment]
Should probably remove `runCommandInStdenvLocal` due to a similar rationale as `runCommandCCLocal`.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
I’d say this is the point where we should introduce a generic version that takes an attrset of these options, instead of creating a matrix of everything.

`runCommandWith { stdenv = … }` …

File: pkgs/build-support/trivial-builders.nix [hunk 1]
```diff
@@ -42,13 +27,56 @@ rec {
...
+    { stdenv ? stdenv
+    , runLocal ? false
+    , derivationArgs ? {}
+    , name
+    }: buildCommand:
+    stdenv.mkDerivation ({
+      name = lib.strings.sanitizeDerivationName name;
+      inherit buildCommand;
+      passAsFile = [ "buildCommand" ]
+        ++ (derivationArgs.passAsFile or []);
```
Comment (Profpatsch) [inline pkgs/build-support/trivial-builders.nix:71]: I’d do this change in a second commit.

Comment (Profpatsch) [reply pkgs/build-support/trivial-builders.nix:71]: That way the first commit is a pure refactor for the existing commands, and the second changes behavior.

**Review by Profpatsch — CHANGES_REQUESTED** [review]
Looks good!

**Comment by sternenseemann** [comment]
New stuff that works:

```nix
with (import ./. {});

runCommandWith {
  # llvm stdenv rebuild on master rn
  stdenv = (import <nixpkgs> {}).llvmPackages.stdenv;
  name = "lol";
} "echo $CC > $out"
# $out will be "clang"
```

```nix
with (import ./. {});

runCommandWith {
  name = "lol";
  derivationArgs = {
    passAsFile = [ "foo" ];
    foo = "bar";
  };
} "mv $fooPath $out"
# $out will be "bar"
```

File: pkgs/build-support/trivial-builders.nix [hunk 2]
```diff
@@ -43,13 +27,60 @@ rec {
...
+    stdenv = stdenvNoCC;
+    runLocal = true;
+    inherit name;
+    derivationArgs = env;
+  };
 
-  runCommandCC = runCommand' false stdenv;
+  runCommandCC = name: env: runCommandWith {
+    runLocal = false;
+    inherit stdenv;
```
Comment (Profpatsch) [inline pkgs/build-support/trivial-builders.nix:45]: It feels a bit hard to follow to me to inherit from the file scope, I’d say `stdenv = stdenv` here is okay.

File: pkgs/build-support/trivial-builders.nix [hunk 3]
```diff
@@ -43,13 +27,60 @@ rec {
...
+   */
+  runCommandWith =
+    let
+      # prevent infinite recursion with default no stdenv given
+      defaultStdenv = stdenv;
+    in
+    { stdenv ? defaultStdenv
+    , runLocal ? false
+    , derivationArgs ? {}
+    , name
```
Comment (Profpatsch) [inline pkgs/build-support/trivial-builders.nix:69]: document these with comments I’d say

**Review by Profpatsch — APPROVED** [review]
two small comments, then LGTM

File: pkgs/build-support/trivial-builders.nix (hunk 2 above)
Comment (sternenseemann) [reply pkgs/build-support/trivial-builders.nix:45]: It's a shame can't have `pkgs` as an argument to this file, really.

**Comment by sternenseemann** [comment]
@ofborg eval

**Comment by Profpatsch** [comment]
I’d say we can merge.
